### PR TITLE
apple_dns: Add DNS query trace

### PR DESCRIPTION
**Pull request by fredyw**
This PR adds trace info in the DNS resolution details by setting the `envoy.enable_dfp_dns_trace` runtime flag to true. This can be useful for debugging issues, such as debugging resolve timeout issues.

Risk Level: low (only enabled with `enable_dfp_dns_trace` runtime flag)
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: dynamic forward proxy, apple dns

### CODE REVIEW

File: source/extensions/network/dns_resolver/apple/apple_dns_impl.cc [hunk 1]
```diff
@@ -347,6 +362,7 @@ void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
     pending_response_.v4_responses_.clear();
     pending_response_.v6_responses_.clear();
 
+    addTrace(static_cast<uint8_t>(AppleDnsTrace::Failed));
```
Comment (alyssawilk) [inline source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:365]: do we want to differentiate this failure mode?  AFIK it makes sense to

File: source/extensions/network/dns_resolver/apple/apple_dns_impl.cc [hunk 2]
```diff
@@ -189,6 +203,7 @@ void AppleDnsResolverImpl::PendingResolution::onEventCallback(uint32_t events) {
     // Therefore, finish resolving with an error.
     pending_response_.status_ = ResolutionStatus::Failure;
     pending_response_.details_ = absl::StrCat("apple_dns_error_", error);
+    addTrace(static_cast<uint8_t>(AppleDnsTrace::Failed));
     finishResolve();
```
Comment (alyssawilk) [inline source/extensions/network/dns_resolver/apple/apple_dns_impl.cc:207]: Could we make finishResolve take the argument and add trace there?

**Review by alyssawilk — COMMENTED** [review]
nice change!  two small nits
/wait

File: source/extensions/network/dns_resolver/apple/apple_dns_impl.cc (hunk 2 above)
Comment (fredyw) [reply source/extensions/network/dns_resolver/apple/apple_dns_impl.cc]: Done.

File: source/extensions/network/dns_resolver/apple/apple_dns_impl.cc (hunk 1 above)
Comment (fredyw) [reply source/extensions/network/dns_resolver/apple/apple_dns_impl.cc]: There can be a lot of different possible failures and it's probably not easy to enumerate all of them. The [getaddrinfo](https://github.com/envoyproxy/envoy/blob/0e1e3d4b724e8d108815589e09df31132faed71e/source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc#L213) implementation also doesn't differentiate between different failures, except for the cases where we have special logic.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**